infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -639,6 +639,11 @@ public void connect(ClusterDescriptor remoteClusterDescriptor, LogReplicationSes
         log.info("Connect asynchronously to remote cluster {} and session {} ", remoteClusterDescriptor.getClusterId(),
                 session);
 
+        if (incomingSession.contains(session)) {
```

Comment:
I agree with adding the RemoteSourceLeadershipManager here.  Can you remove it from here -
https://github.com/CorfuDB/CorfuDB/blob/log_replication_v2/infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java#L343

Also, please add a comment similar to 'If connection initiator and Sink, add the Source Leadership Manager for this session'.

Reply:
Thanks for catching this. It's another part I missed during rebase. Previously this map has a race condition that the `connectionAsync` thread could access it before it's even updated. So Shama moved this logic to `SessionManager`'s refresh method and it's guaranteed that the map is updated first. It could be removed from here then.